### PR TITLE
CarPlay refactoring and bugfixing.

### PR DESCRIPTION
**Pull request by MaximAlien**
### Description

PR brings some improvements and changes, which are needed for https://github.com/mapbox/mapbox-navigation-ios-examples/pull/134.

Also fixes:
- #3220
- ~~#3224~~

### CODE REVIEW

File: Sources/MapboxNavigation/CarPlayMapViewController.swift [hunk 1]
```diff
@@ -306,10 +306,9 @@ extension CarPlayMapViewController: StyleManagerDelegate {
     }
     
     public func styleManager(_ styleManager: StyleManager, didApply style: Style) {
-        let styleURL = style.previewMapStyleURL
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/CarPlayMapViewController.swift:309]: Fixes #3224. There should be an ability to change `MapView` style URL similarly to iOS.

File: Sources/MapboxNavigation/CarPlayManagerDelegate.swift [hunk 2]
```diff
@@ -161,6 +161,38 @@ public protocol CarPlayManagerDelegate: AnyObject, UnimplementedLogging {
...
+                        templateWillDisappear template: CPTemplate,
+                        animated: Bool)
+    
+    /**
+     Delegate method, which allows to observe when template presented by the `CarPlayManager` did
+     disappear from the screen.
+     */
+    func carPlayManager(_ carPlayManager: CarPlayManager,
+                        templateDidDisappear template: CPTemplate,
+                        animated: Bool)
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/CarPlayManagerDelegate.swift:195]: Delegate methods, which were added in scope of #3220.

File: Sources/MapboxNavigation/CarPlayManager.swift [hunk 3]
```diff
@@ -417,21 +420,36 @@ extension CarPlayManager: CPInterfaceControllerDelegate {
...
+        delegate?.carPlayManager(self, templateDidDisappear: template, animated: animated)
+    }
 }
 
 @available(iOS 12.0, *)
 extension CarPlayManager {
     
+    /**
+     Allows to preview routes for specific destination `Waypoint` by calculating the routes asynchronously
+     using the [Mapbox Directions API](https://www.mapbox.com/api-documentation/navigation/#directions).
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManager.swift:443]:
```suggestion
     Calculates routes to the given destination using the [Mapbox Directions API](https://www.mapbox.com/api-documentation/navigation/#directions) and previews them on a map.
     .
```

File: Sources/MapboxNavigation/CarPlayManager.swift [hunk 4]
```diff
@@ -451,29 +469,46 @@ extension CarPlayManager {
...
+     - parameter waypoints: A list of `Waypoint` objects.
+     - parameter completionHandler: A closure to be executed when the calculation completes.
+     */
     public func previewRoutes(between waypoints: [Waypoint], completionHandler: @escaping CompletionHandler) {
         let options = NavigationRouteOptions(waypoints: waypoints)
         previewRoutes(for: options, completionHandler: completionHandler)
     }
     
+    /**
+     Allows to preview routes based on criteria, which is specified in `RouteOptions` object.
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManager.swift:484]:
```suggestion
     Calculates routes satisfying the given options using the [Mapbox Directions API](https://www.mapbox.com/api-documentation/navigation/#directions) and previews them on a map.
```

File: Sources/MapboxNavigation/CarPlayManager.swift [hunk 5]
```diff
@@ -451,29 +469,46 @@ extension CarPlayManager {
...
+     
+     - parameter routeOptions: A `RouteOptions` object, which specifies the criteria for results
+     returned by the Mapbox Directions API.
+     - parameter completionHandler: A closure to be executed when the calculation completes.
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManager.swift:488]: Would it be more accurate to say that this closure gets executed once the preview is fully visible?

File: Sources/MapboxNavigation/CarPlayManagerDelegate.swift [hunk 6]
```diff
@@ -161,6 +161,38 @@ public protocol CarPlayManagerDelegate: AnyObject, UnimplementedLogging {
                         didAdd finalDestinationAnnotation: PointAnnotation,
                         to parentViewController: UIViewController,
                         pointAnnotationManager: PointAnnotationManager)
+    
+    /**
+     Delegate method, which allows to observe when template presented by the `CarPlayManager` will
+     appear on the screen.
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManagerDelegate.swift:167]:
```suggestion
     Called when a template presented by the `CarPlayManager` is about to appear on the screen.
```

File: Sources/MapboxNavigation/CarPlayMapViewController.swift [hunk 7]
```diff
@@ -306,10 +306,9 @@ extension CarPlayMapViewController: StyleManagerDelegate {
...
         let mapboxMapStyle = navigationMapView.mapView.mapboxMap.style
         if mapboxMapStyle.uri?.rawValue != style.mapStyleURL.absoluteString {
-            mapboxMapStyle.uri = StyleURI(url: styleURL)
+            mapboxMapStyle.uri = StyleURI(url: style.mapStyleURL)
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayMapViewController.swift:311]: CarPlayMapViewController appears in the browsing and previewing activities. These activities should use `previewMapStyleURL`, not `mapStyleURL`, which is intended for the navigating activity. Although the default Navigation Day and Navigation Night styles no longer distinguish between previewing and turn-by-turn navigation with two sets of styles, it remains possible and possibly desirable for a developer to keep the two scenarios separate.

**Review by 1ec5 — COMMENTED** [review]
Mostly documentation nits, but the removal of `previewMapStyleURL` usage would be a regression. Maybe I’m misunderstanding the impetus for that change though.

File: Sources/MapboxNavigation/CarPlayMapViewController.swift (hunk 7 above)
Comment (MaximAlien) [reply Sources/MapboxNavigation/CarPlayMapViewController.swift]: Ah, I've completely missed that. Reverted to the previous state.

File: Sources/MapboxNavigation/CarPlayManager.swift (hunk 5 above)
Comment (MaximAlien) [reply Sources/MapboxNavigation/CarPlayManager.swift:487]: Yep, agree.

File: Sources/MapboxNavigation/CarPlayManagerDelegate.swift [hunk 8]
```diff
@@ -161,6 +161,37 @@ public protocol CarPlayManagerDelegate: AnyObject, UnimplementedLogging {
...
+    /**
+     Called when a template presented by the `CarPlayManager` is about to appear on the screen.
+     */
+    func carPlayManager(_ carPlayManager: CarPlayManager,
+                        templateWillAppear template: CPTemplate,
+                        animated: Bool)
+    
+    /**
+     Delegate method, which allows to observe when template presented by the `CarPlayManager` did
+     appear on the screen.
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManagerDelegate.swift:174]:
```suggestion
     Called when a template presented by the `CarPlayManager` has finished appearing on the screen.
```

File: Sources/MapboxNavigation/CarPlayManagerDelegate.swift [hunk 9]
```diff
@@ -161,6 +161,37 @@ public protocol CarPlayManagerDelegate: AnyObject, UnimplementedLogging {
...
+     */
+    func carPlayManager(_ carPlayManager: CarPlayManager,
+                        templateDidAppear template: CPTemplate,
+                        animated: Bool)
+    
+    /**
+     Delegate method, which allows to observe when template presented by the `CarPlayManager` will
+     disappear from the screen.
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManagerDelegate.swift:182]:
```suggestion
     Called when a template presented by the `CarPlayManager` is about to disappear from the screen.
```

File: Sources/MapboxNavigation/CarPlayManagerDelegate.swift [hunk 10]
```diff
@@ -161,6 +161,37 @@ public protocol CarPlayManagerDelegate: AnyObject, UnimplementedLogging {
...
+     */
+    func carPlayManager(_ carPlayManager: CarPlayManager,
+                        templateWillDisappear template: CPTemplate,
+                        animated: Bool)
+    
+    /**
+     Delegate method, which allows to observe when template presented by the `CarPlayManager` did
+     disappear from the screen.
```
Comment (1ec5) [inline Sources/MapboxNavigation/CarPlayManagerDelegate.swift:190]:
```suggestion
     Called when a template presented by the `CarPlayManager` has finished disappearing from the screen.
```

**Review by 1ec5 — APPROVED** [review]
Sorry, these suggestions didn’t take for some reason. Looks good overall.

**Pull request merged**